index/scorch/snapshot_index.go

```diff
@@ -507,11 +523,29 @@ func (i *IndexSnapshot) documentVisitFieldTerms(id index.IndexInternalID,
 	ss := i.segment[segmentIndex]
 
 	if zaps, ok := ss.segment.(segment.DocumentFieldTermVisitable); ok {
-		// get the list of doc value persisted fields
-		pFields, err := zaps.VisitableDocValueFields()
-		if err != nil {
-			return nil, err
+		pFields := ss.cachedDocs.getPersistedFields()
+		// if all fields are dv presisted
+		if len(pFields) > 0 && checkSubset(pFields, fields) {
+			return zaps.VisitDocumentFieldTerms(localDocNum, fields, visitor, dvs)
+		}
+
+		cFields := ss.cachedDocs.getCachedFields()
+		// if all fields are already in dvCache or persisted
+		if len(cFields) > 0 && checkSubset(append(pFields, cFields...), fields) {
```

Comment:
this approaching to union'ing the pFields and cFields is unsafe, as the underlying array of the pFields slice is shared across multiple threads, even though it was probably intended to be immutable.  The append(), though, may mutate the input slice's underlying backing array.

Reply:
i belive this comment was already made by sreekanth an in our morning review, not sure what happened to that